runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/FnApiControlClientPoolService.java

```diff
@@ -79,7 +76,7 @@ public static FnApiControlClientPoolService offeringClientsToPool(
       // discarded, which should be performed by a call to #shutdownNow. The remote caller must be
       // able to handle an unexpectedly terminated connection.
       vendedClients.add(newClient);
```

Comment:
<!--thread_id:cc_182896326_t; commit:1fd121da1417624b3b84f0300648251da64b9cb5; resolved:0-->
<!--section:context-quote-->
> **bsidhom** wrote:
> Ack. I've added synchronization around `vendedClients` itself. It may be worth synchronizing the entire close method now that we're synchronizing over the bulk of its work. @tgroh What do you think?

<!--section:body-->
Actually, I just realized that the list is already backed by a CopyOnWriteArrayList and we're using the atomic boolean in close in conjunction. However, if we want to protect against incoming clients getting leaked after a call to close, we should do away with the atomic/COW and just use synchronization where necessary.

Another note about leak prevention: While using the vended client list prevents leakages of unclosed but invalid InstructionRequestHandlers, we're also telling ControlClientPool consumers that they're responsible for cleanup. So a consumer may terminate a control client and remove all references to it, expecting it to be GC'd. However, clients will never be garbage collected until the control server is shut down.

It seems plausible (depending on the runner) to have a long-lived control server that is shared between jobs and that controls many different environments during its lifespan. If we want to deal with that case, we may need to hold only `WeakReference`s to clients. @tgroh Do we care about this scenario? Are control servers even allowed to be reused between jobs?